File: spring-web-java/src/main/java/spring/web/java/global/filter/JwtAuthenticationFilter.java
package spring.web.java.global.filter;

import java.io.IOException;
import java.util.List;

import org.springframework.http.HttpHeaders;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.authority.SimpleGrantedAuthority;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.util.StringUtils;
import org.springframework.web.filter.OncePerRequestFilter;

import io.jsonwebtoken.Claims;
import jakarta.servlet.FilterChain;
import jakarta.servlet.ServletException;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;
import lombok.RequiredArgsConstructor;
import spring.web.java.global.common.JwtTokenProvider;

@RequiredArgsConstructor
public class JwtAuthenticationFilter extends OncePerRequestFilter {

	private final JwtTokenProvider jwtTokenProvider;

	@Override
	protected void doFilterInternal(
		HttpServletRequest request,
		HttpServletResponse response,
		FilterChain filterChain
	) throws ServletException, IOException {
		String token = getToken(request);

		if (StringUtils.hasText(token)) {
			Claims claims = jwtTokenProvider.parseAccessToken(token);
			Authentication authentication = new UsernamePasswordAuthenticationToken(
				claims.get("memberId"),
				token,
				List.of(new SimpleGrantedAuthority(String.valueOf(claims.get("role"))))
			);
			SecurityContextHolder.getContext().setAuthentication(authentication);
		}

		filterChain.doFilter(request, response);
	}

	public String getToken(HttpServletRequest request) {
		String authorization = request.getHeader(HttpHeaders.AUTHORIZATION);

		if (StringUtils.hasText(authorization) && authorization.startsWith("Bearer")) {
			return authorization.replace("Bearer ", "");
		}

		return null;
	}
}

File: spring-web-kotlin/src/main/kotlin/spring/web/kotlin/config/security/JwtAuthenticationFilter.kt
package spring.web.kotlin.config.security

import jakarta.servlet.FilterChain
import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletResponse
import org.springframework.http.HttpHeaders
import org.springframework.security.authentication.AuthenticationManager
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken
import org.springframework.security.core.context.SecurityContextHolder
import org.springframework.security.web.authentication.www.BasicAuthenticationFilter

class JwtAuthenticationFilter(
    private val authenticationManager: AuthenticationManager,
    private val jwtTokenProvider: JwtTokenProvider
) : BasicAuthenticationFilter(authenticationManager) {
    override fun doFilterInternal(request: HttpServletRequest, response: HttpServletResponse, chain: FilterChain) {

        getAccessToken(request)
            ?.let(jwtTokenProvider::parseAccessToken)
            ?.let(::UserDetailsImpl)
            ?.let { UsernamePasswordAuthenticationToken(it, it.password, it.authorities) }
            ?.let { SecurityContextHolder.getContext().authentication = it }

        chain.doFilter(request, response)
    }

    private fun getAccessToken(request: HttpServletRequest) = request.getHeader(HttpHeaders.AUTHORIZATION)
        .takeIf { it.startsWith("Bearer") }
        ?.replace("Bearer ", "")
}